### ISSUE TITLE
Checkout remote branch should allow entering branch name without "origin/"

### ISSUE THREAD

**Issue by jjcole** [gitextensions/gitextensions#6423]
## Current behaviour

Copying branch name without "origin/" doesn't work when checking out remote branch.

## Expected behaviour

Paste or type in name of branch without "origin/", drop-down should appear allowing you to press enter or click on the correct branch.

## Steps to reproduce

1. CTRL-. to Checkout branch
2. Select "Remote branch"
3. Enter branch name without "origin/"

## Screenshots

![image](https://user-images.githubusercontent.com/581226/55260325-77a0fd00-523e-11e9-9a91-c7442b4507a0.png)

![image](https://user-images.githubusercontent.com/581226/55260538-0d3c8c80-523f-11e9-898e-3ed42e58a649.png)

## Did this work in previous version of GitExtensions

Not that I know of.

## Environment

Git Extensions 3.0.2.5232
Build 23b6f51
Git 2.20.1.windows.1
Microsoft Windows NT 10.0.14393.0
.NET Framework 4.7.3394.0
DPI 96dpi (no scaling)

**Comment by illfated** [comment]
Hmm, that does not look practical to me (unless I have misunderstood your point completely).
Let's say that I fork gitextensions/gitextensions and clone it to my local computer. 
I would then rename "upstream" to "gitextensions" to match the name of the repository owner. 
Next, I would rename my "origin" fork to my username, since its path would be **illfated**/gitextensions. 
My cloned repository will then be only "gitextensions" on my computer.

To distinguish between the local repository and the remotes, I presume it needs either 
a choice between the names I have assigned to the remotes (gitextensions or illfated), 
or keep using "upstream" & "origin" as the preceding pointer.

**Comment by gerhardol** [comment]
This would require to match to every possible remote. Completion works, matching the remotename, so you will have to type o-right-arrow.

**Comment by jjcole** [comment]
Maybe I'm just misunderstanding something since I only have one remote, "origin", in my repos. But I can click the drop-down and I see this:

![image](https://user-images.githubusercontent.com/581226/55270624-6aeccb00-5277-11e9-8612-52e70cbc9620.png)

I'm just suggesting that typing in <branch> should find and bring up origin/<branch> if it would already be visible in that drop-down that I opened.

I imagine if that drop-down contained branches for multiple remotes, that it would potentially bring up remote1/<branch> and remote2/<branch> if a branch with that name existed in multiple remotes.

**Comment by vbjay** [comment]
Type of then hit right arrow. It's likely to have multiple remotes. If you
had 3 remotes with the branch mywork which one should we match. Instead we
let you first filter by remote and then see the branches in that remote. As
you type it will filter down. I really don't see a problem.

**Comment by illfated** [comment]
I agree. Git Extensions can't know if the remote you are looking for belongs to an `upstream` remote or an `origin` remote, so it needs to have one of those 2 defined before entering the name of the remote.

**Comment by gerhardol** [comment]
An enhancement request could be worded as:
Completion should match not just on the start of the string but on all directories.

So if the following remote branches exists:
remote1/feature/my-change1
remote1/bugfix/fix2
remote2/feature/mychange3

Typing "my" will list branch 1 and 3

**Comment by jjcole** [comment]
@vbjay If there are multiple branches over multiple remotes then all matching options should be shown in the drop-down, the same way it already does if you search for a string that exists at the beginning of multiple branches, like this:

![image](https://user-images.githubusercontent.com/581226/55277623-c26f5300-52d8-11e9-964c-bd34af7fd35a.png)

@gerhardol Exactly, I think that's the true issue. Besides what I'm suggesting, you should also be able to search "change" and it should bring up remote1/feature/my-change1 AND remote2/feature/mychange3 in the drop-down.

**Comment by BOBJohnson** [comment]
I would like to see this feature as well. I only ever work with one remote. Either the ability to match any branch regardless of remote name as it filters the list / suggestions or the ability in the 'Configure Remotes' to define a default remote for a given project so that when I go to Checkout -> Remote branch, the text already has 'defaultremote/' and the keyboard focus is already after the slash.

**Comment by vbjay** [comment]
Thej.do it right. Checkout remote branch and create a local branch. Copying the branch name means  copying the  current ref. Depending on which option you seke t in checkout of a 4emote branch controls what ref is checked out.

**Comment by BOBJohnson** [comment]
@vbjay I did not understand your response.

However after poking around in the code here and doing some research I have determined the main limitation: the winforms autocomplete functionality only supports 'starts with' and not anything custom or fancy like 'contains'. You are then left with a couple possibilities:

A) Write your own combobox control that implements it's own custom filtering on text input that then displays an attached ListBox as a popup. This seems like a lot of work for a nuisance feature.
B) Purchase one. I couldn't find any free ones (that doesn't mean there isn't one - I just couldn't find one). That's not really feasible here either. An example of one though is Telerik's RadDropDownList (https://docs.telerik.com/devtools/winforms/controls/dropdown-listcontrol-and-checkeddropdownlist/dropdownlist/overview) which has a 'Contains' mode, but also supports you providing your own matching algorithm.

Either which way, getting 'contains' to work from what research I did seemed like a nightmare on winforms.

Now I don't know about creating Visual Studio extensions, but if you are amenable to WPF, there's some free solutions available that might be viable.
